### PR TITLE
[c10d][jit] initial torchbind bindings for ProcessGroupNCCL

### PR DESCRIPTION
**Pull request by wanchaol**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#42944 [c10d][jit] initial torchbind bindings for ProcessGroupNCCL**

This PR introduces initial custom class bindings for ProcessGroup, and
make it work specifically for ProcessGroupNCCL.

We didn't make the API exactly the same with python pybind bindings, as
there are some issues with TorchBind to support some argument types and
we will resolve that separately to make these two stays the same. And
eventually, when we have a good TorchBind coverage, we will switch all
bindings in c10d TorchBind

Differential Revision: [D23228682](https://our.internmc.facebook.com/intern/diff/D23228682)

### CODE REVIEW

File: torch/lib/c10d/ProcessGroup.cpp [hunk 1]
```diff
@@ -1,9 +1,34 @@
...
+    .def("source_rank", &::c10d::ProcessGroup::Work::sourceRank)
+    .def("synchronize", &::c10d::ProcessGroup::Work::synchronize);
+
+// Torchbind AllToAllOptions to make it available in TorchScript
+// NOTE: we only exposed the constructor, and didn't expose the
+// attributs to Python/TS yet because of TorchBind limitation
+// TODO: remove pybind and universally use TorchBind when TorchBind
+// could handle std::chrono like pybind
+static auto allToAllOptions =
+  torch::class_<::c10d::AllToAllOptions>("dist_c10d", "AllToAllOption")
```
Comment (gmagogsfm) [inline torch/lib/c10d/ProcessGroup.cpp:25]: typo? AlltToAllOption*s*

File: torch/lib/c10d/ProcessGroup.cpp [hunk 2]
```diff
@@ -1,9 +1,34 @@
...
+    .def(torch::init<>());
+
+static auto processGroup =
```
Comment (gmagogsfm) [inline torch/lib/c10d/ProcessGroup.cpp:28]: I moved a bit further after adding AllToAllOptions TorchBind. 

But then hit issue where ProcessGroup doesn't have "alltoall" defined. After adding "alltoall" binding here, TorchBind complains about unable to handle non-const reference:
https://fburl.com/paste/nqoiu8zt

File: test/distributed/test_jit_c10d.py [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+        opts = c10d.ProcessGroupGloo.Options()
+        opts.devices = [c10d.ProcessGroupGloo.create_device(interface=LOOPBACK)]
+        opts.timeout = 5.0
+        opts.threads = threads
+        return opts
+
+    def test_jit_process_group(self):
+        @torch.jit.script
+        def test_create_store(filename: str, world_size: int) -> torch.classes.dist_c10d.FileStore:
+            store = torch.classes.dist_c10d.FileStore(filename, world_size)
```
Comment (pritamdamania87) [inline test/distributed/test_jit_c10d.py:44]: Is this what the user facing API would look like with torchbind? Is there a way to get around this and use `torch.distributed.FileStore`? The reason being that this is kind of an API change for the user if they would like to use the same code under TorchScript.

Comment (wanchaol) [reply test/distributed/test_jit_c10d.py:44]: ultimately we would like to make `torch.distributed.FileStore = torch.classes.dist_c10d.FileStore`, this way the user could just reuse the existing API either in python and in TorchScript. But this currently is not possible due to various TorchBind limitations (e.g. [support `std::chrono:milisecond` as argument](https://github.com/pytorch/pytorch/issues/42948), [install doc via TorchBind](https://github.com/pytorch/pytorch/issues/42951), etc.) So I am just purely testing it here to make sure the TorchBind part works fine, and we will improve the TorchBind to cover those missing features before fully move to TorchBind

**Review by gmagogsfm — APPROVED** [review]
Just a few minor comments discussed offline. LGTM.

File: torch/csrc/distributed/c10d/init.cpp [hunk 4]
```diff
@@ -1223,6 +1223,83 @@ that adds a prefix to each key inserted to the store.
 
 #undef PROCESS_GROUP_DEPRECATION_WARNING
 
+// NOTE: Below are TorchBind bindings for c10d, these bindings will
+// live together with those pybind11 bindings above until we resolve
+// all the TorchBind issues and merge these two together. we shouldn't
+// document this until we finish the migration.
+
+static const auto StoreTorchBind =
+    torch::class_<::c10d::Store>("dist_c10d", "Store");
```
Comment (mrshenli) [inline torch/csrc/distributed/c10d/init.cpp:1232]: (not requesting for change)

If `dist_c10d` is what users will see when this is done, let's try to come up with a better name for it. Maybe, sth like `torch.classes.distributed` or `torch.classes.distributed.comm`, etc.

File: torch/csrc/distributed/c10d/init.cpp [hunk 5]
```diff
@@ -1223,6 +1223,83 @@ that adds a prefix to each key inserted to the store.
...
+            })
+        .def("result", &::c10d::ProcessGroup::Work::result);
+
+static const auto ProcessGroupTorchBind =
+    torch::class_<::c10d::ProcessGroup>("dist_c10d", "ProcessGroup");
+
+#ifdef USE_C10D_NCCL
+
+// XXX: Ideally the Options of ProcessGroupNCCL should be
+// binded using `def_readwrite` like in pybind11, but we
```
Comment (mrshenli) [inline torch/csrc/distributed/c10d/init.cpp:1266]: typo: binded -> bound

File: test/distributed/test_jit_c10d.py [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+)
+class ProcessGroupNCCLJitTest(TestCase):
+    MAIN_PROCESS_RANK = 0
+
+    def setUp(self):
+        self.rank = self.MAIN_PROCESS_RANK
+        self.world_size = 1
+        self.file = tempfile.NamedTemporaryFile(delete=False)
+        self.num_gpus = torch.cuda.device_count()
+        if self.num_gpus < 2:
```
Comment (mrshenli) [inline test/distributed/test_jit_c10d.py:37]: hmm, doesn't seem the test need two GPUs?

File: test/distributed/test_jit_c10d.py [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+            output: torch.Tensor,
+            input: torch.Tensor
+        ):
+            output_split_sizes: List[int] = []
+            input_split_sizes: List[int] = []
+            work = pg.alltoall_base(output, input, output_split_sizes, input_split_sizes)
+            work.wait()
+            return work.result()
+
+        run_pg_nccl_alltoall(nccl_pg, output, input)
```
Comment (mrshenli) [inline test/distributed/test_jit_c10d.py:73]: shall we apply `self.assert*` to check the result?

**Review by mrshenli — COMMENTED** [review]
qq, what's the benefit of using torchbind compared to pybind?

**Review by mrshenli — APPROVED** [review]

File: torch/csrc/distributed/c10d/init.cpp (hunk 4 above)
Comment (wanchaol) [reply torch/csrc/distributed/c10d/init.cpp:1232]: yes I agree with that, the naming is bad. That's something torchbind didn't support yet, we will be adding that support soon and I will switch the names over in follow up PRs.

**Pull request closed without merging**